connectors/cassandra/schema.go

```diff
@@ -37,17 +55,197 @@ func (c *Connector) TruncateScope(ctx context.Context, scope string) error {
+			return fmt.Errorf("Table %q clustering key mismatch (column %d should be %q)", ed.Name, i+1, ck.Name)
+		}
+	}
+
+	// Check each column
+	for _, col := range ed.Columns {
+		_, ok := md.Columns[col.Name]
+		if !ok {
+			schemaErrors.MissingColumns = append(schemaErrors.MissingColumns, MissingColumn{Column: *col, Tablename: ed.Name})
+		}
+		// TODO: check column type
+	}
+
+	return nil
+
+}
+
+// CheckSchema verifies that the schema passed in the registered entities matches the database
 func (c *Connector) CheckSchema(ctx context.Context, scope string, namePrefix string, ed []*dosa.EntityDefinition) (int32, error) {
-	return 0, nil
+	schemaErrors := new(RepairableSchemaMismatchError)
+
+	// TODO: unfortunately, gocql doesn't have a way to pass the context to this operation :(
+	km, err := c.Session.KeyspaceMetadata(c.KsMapper.Keyspace(scope, namePrefix))
+	if err != nil {
+		return 0, err
+	}
+	for _, ed := range ed {
+		tableMetadata, ok := km.Tables[ed.Name]
+		if !ok {
+			schemaErrors.MissingTables = append(schemaErrors.MissingTables, ed.Name)
+			continue
+		}
+		if err := compareStructToSchema(ed, tableMetadata, schemaErrors); err != nil {
+			return 0, err
+		}
+
+	}
+	if schemaErrors.HasMissing() {
+		return 0, schemaErrors
+	}
+	return int32(1), nil
 }
 
-// UpsertSchema is not implemented
+// UpsertSchema checks the schema and then updates it
 func (c *Connector) UpsertSchema(ctx context.Context, scope string, namePrefix string, ed []*dosa.EntityDefinition) (*dosa.SchemaStatus, error) {
-	panic("next diff")
+	sr, err := c.CheckSchema(ctx, scope, namePrefix, ed)
+	if repairs, ok := err.(*RepairableSchemaMismatchError); ok {
+		for _, table := range repairs.MissingTables {
+			for _, def := range ed {
+				if def.Name == table {
+					err = c.createTable(c.KsMapper.Keyspace(scope, namePrefix), def)
+					if err != nil {
+						return nil, errors.Wrapf(err, "Unable to create table %q", def.Name)
+					}
+				}
+			}
+		}
+		for _, col := range repairs.MissingColumns {
+			if err := c.alterTable(c.KsMapper.Keyspace(scope, namePrefix), col.Tablename, col.Column); err != nil {
+				return nil, errors.Wrapf(err, "Unable to add column %q to table %q", col.Column.Name, col.Tablename)
+			}
+		}
+	}
+	return &dosa.SchemaStatus{Version: sr, Status: "ready"}, err
+}
+
+func (c *Connector) alterTable(keyspace, table string, col dosa.ColumnDefinition) error {
+	return c.Session.Query(alterTableString(keyspace, table, col)).Exec()
+}
+
+func alterTableString(keyspace, table string, col dosa.ColumnDefinition) string {
+	cts := &bytes.Buffer{}
+	fmt.Fprintf(cts, `ALTER TABLE "%s"."%s" ( ADD "%s" %s )`, keyspace, table, col.Name, cassandraType(col.Type))
+	return cts.String()
+}
+
+func (c *Connector) createTable(keyspace string, ed *dosa.EntityDefinition) error {
+	return c.Session.Query(createTableString(keyspace, ed)).Exec()
+}
+
+func createTableString(keyspace string, ed *dosa.EntityDefinition) string {
```

Comment:
When generating CQL for all of our crud and queries we use go's templating framework, why not also use the templating here? I think it will make the code way easier to read.